File: 面经150p2/30.串联所有单词的子串.cpp
/*
 * @lc app=leetcode.cn id=30 lang=cpp
 *
 * [30] 串联所有单词的子串
 */

// @lc code=start
#pragma once
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
#pragma GCC optimize("Ofast,no-stack-protector,unroll-loops")
#define ALL(v) v.begin(), v.end()
#define For(i, _) for (int i = 0, i##end = _; i < i##end; ++i)
#define FOR(i, _, __) for (int i = _, i##end = __; i < i##end; ++i)
#define Rep(i, _) for (int i = (_); i >= 0; --i)
#define REP(i, __, _) for (int i = (__), i##end = _; i >= i##end; --i)
typedef long long ll;
typedef unsigned long long ull;
#define V vector
#define pb push_back
#define pf push_front
#define qb pop_back
#define qf pop_front
#define eb emplace_back
typedef pair<int, int> pii;
typedef pair<ll, int> pli;
#define fi first
#define se second
const int dir[4][2] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
const int inf = 0x3f3f3f3f;
const ll infl = 0x3f3f3f3f3f3f3f3fll;
const int mod = 1e9 + 7;
template <class T>
inline bool ckmin(T &x, const T &y) { return x > y ? (x = y, true) : false; }
template <class T>
inline bool ckmax(T &x, const T &y) { return x < y ? (x = y, true) : false; }
int init = []()
{ ios::sync_with_stdio(false); cin.tie(nullptr); return 0; }();

class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        int n = s.size();
        int idx = 0;
        int sz = words[0].size();
        int sz_total = sz * words.size();
        V<int> res;
        unordered_map<string, int> hash;
        for (auto& w : words) hash[w] ++;
        FOR(i, 0, min(sz, n)) {
            int v = i;
            unordered_map<string, int> tmp;
            while (v + sz_total <= n) {
                if (v == i) {
                    int u = i;
                    while (u < v + sz_total) {
                        tmp[s.substr(u, sz)] += 1;
                        u += sz;
                    }
                }
                else {
                    string l = s.substr(v - sz, sz);
                    tmp[l] -= 1;
                    if (tmp[l] == 0) tmp.erase(l);
                    string r = s.substr(v + sz_total - sz, sz);
                    tmp[r] += 1;
                }
                if (tmp == hash) res.pb(v);
                v += sz;
            }
        }
        return res;
    }
};
// @lc code=end

File: 面经150/面经150p1/3.滑动窗口/30.串联所有单词的子串.py
#
# @lc app=leetcode.cn id=30 lang=python3
#
# [30] 串联所有单词的子串
#
from mytools import *
# @lc code=start
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        hash = defaultdict(int)
        for st in words: hash[st] += 1
        sz1 = len(words); sz2 = len(words[0])
        sz = sz1 * sz2
        n = len(s)
        res = []
        for offset in range(sz2):
            i = offset
            while i + sz <= n:
                if i == offset:
                    tmp = defaultdict(int)
                    j = i
                    while j < i + sz:
                        tmp[s[j:j + sz2]] += 1
                        j += sz2
                else:
                    l = s[i - sz2:i]
                    tmp[l] -= 1
                    if tmp[l] == 0: del tmp[l]
                    nw = s[i + sz - sz2:i + sz]
                    tmp[nw] += 1
                if tmp == hash: res.append(i)
                i += sz2
        return res
# @lc code=end

class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        hash = defaultdict(int)
        for st in words: hash[st] += 1
        sz1 = len(words); sz2 = len(words[0])
        sz = sz1 * sz2
        n = len(s)
        res = []
        i = j = 0
        while i < n:
            tmp = defaultdict(int)
            if i + sz > n: break
            j = i
            while j < i + sz:
                s1 = ""
                s1 = s[j:j + sz2]
                tmp[s1] += 1
                j += sz2
            if hash == tmp: res.append(i)
            i += 1
        return res
            